Data/ByteString/Lazy.hs

```diff
@@ -1429,19 +1433,34 @@ unzip ls = (pack (List.map fst ls), pack (List.map snd ls))
 -- ---------------------------------------------------------------------
 -- Special lists
 
--- | /O(n)/ Return all initial segments of the given 'ByteString', shortest first.
+-- | Returns all initial segments of the given 'ByteString', shortest first.
 inits :: ByteString -> [ByteString]
-inits = (Empty :) . inits'
-  where inits' Empty        = []
-        inits' (Chunk c cs) = List.map (`Chunk` Empty) (List.drop 1 (S.inits c))
-                           ++ List.map (Chunk c) (inits' cs)
+inits bs = NE.toList $! initsNE bs
 
--- | /O(n)/ Return all final segments of the given 'ByteString', longest first.
+-- | Returns all initial segments of the given 'ByteString', shortest first.
+--
+-- @since 0.11.4.0
+initsNE :: ByteString -> NonEmpty ByteString
+initsNE = (Empty :|) . inits' id
+  where
+    inits' :: (ByteString -> ByteString) -> ByteString -> [ByteString]
+    -- inits' f bs === map f (tail (inits bs))
+    inits' _ Empty = []
+    inits' f (Chunk c@(S.BS x len) cs)
+      = [f (S.BS x n `Chunk` Empty) | n <- [1..len]]
+      ++ inits' (f . Chunk c) cs
+
+-- | /O(n)/ Returns all final segments of the given 'ByteString', longest first.
 tails :: ByteString -> [ByteString]
-tails Empty         = [Empty]
-tails cs@(Chunk c cs')
-  | S.length c == 1 = cs : tails cs'
-  | otherwise       = cs : tails (Chunk (S.unsafeTail c) cs')
+tails bs = NE.toList $! tailsNE bs
+
+-- | /O(n)/ Returns all final segments of the given 'ByteString', longest first.
+--
+-- @since 0.11.4.0
+tailsNE :: ByteString -> NonEmpty ByteString
+tailsNE bs = case  uncons bs  of
+  Nothing -> Empty :| []
+  Just (_, tl) -> bs :| tails tl
```

Comment:
Is it better than ``bs `NE.cons` tailsNE tl``, avoiding mutual recursion?

Reply:
Yes, especially since `cons` is one of the confusingly lazy bits of the `NonEmpty` API; using it here would mean we allocate two extra selector thunks at each loop iteration. GHC easily untangles the mutual recursion; both `tails` and `tailsNE` optimize to small wrappers around the same worker.

Reply:
Might be worth a comment then, otherwise I easily imagine someone "refactoring" it down the road.

Reply:
I added a brief note. But I wasn't quite sure what to write. Is this satisfactory?